astropy/units/core.py

```diff
@@ -2134,20 +2118,17 @@ def __call__(
                     warnings.warn(msg, UnitsWarning)
                 return UnrecognizedUnit(s)
 
-        # can't use units.typing.Complex because of an import loop
-        elif isinstance(s, (int, float, complex, Fraction, np.number)):
+        from .typing import Complex
+
+        if isinstance(s, Complex):
```

Comment:
It may be easy to look up, but it reads weird: for most people, `isinstance(integer, Complex)` will be obviously `False` (as it is for `python`), so a comment is appropriate. Also happy to just keep what we had, or use `numbers.Number` - we're *not* typing here after all (is there really no `typing.Number` equivalent???)

Reply:
> It may be easy to look up, but it reads weird: for most people, `isinstance(integer, Complex)` will be obviously `False` (as it is for `python`), so a comment is appropriate. 

$\mathbb{Z} \subset \mathbb{Q} \subset \mathbb{R} \subset \mathbb{C}$
And in Python:
```python
>>> from numbers import Integral, Rational, Real, Complex
>>> all(issubclass(int, num_set) for num_set in (Integral, Rational, Real, Complex))
True
>>> all(isinstance(0, num_set) for num_set in (Integral, Rational, Real, Complex))
True
```

> Also happy to just keep what we had, or use `numbers.Number` - we're _not_ typing here after all 

We are not type checking _yet_.

> (is there really no `typing.Number` equivalent???)

It is not possible to implement something like `numbers.Number` as a `Protocol` because there are no methods or attributes `Number` is obligated to have, so _everything_ would trivially satisfy the protocol. This fact also screws over the other classes in `numbers` because they [cannot be both implemented as `Protocol` and as subclasses of a non-`Protocol` `Number`](https://peps.python.org/pep-0544/#protocols-subclassing-normal-classes) and they are already subclasses of `Number`. The best Python has figured out is to [declare `int` to be a subtype (not subclass!) of `float` and `float` to be a subtype of `complex`](https://peps.python.org/pep-0484/#the-numeric-tower).